### PR TITLE
User Defined Action Instance Concurrency Limits

### PR DESCRIPTION
**Pull request by bdoyle0182**
## Description
Gives users the ability to configure action concurrency limits on their action with respect to their namespace concurrency limit. This finally allows users to be able to configure some level of fairness within their namespace. See POEM 4 for considerations and design.

## Related issue and scope
- [ ] I opened an issue to propose and discuss this change (#????)

## My changes affect the following components
- [X] API
- [X] Controller
- [ ] Message Bus (e.g., Kafka)
- [ ] Loadbalancer
- [X] Scheduler
- [ ] Invoker
- [ ] Intrinsic actions (e.g., sequences, conductors)
- [ ] Data stores (e.g., CouchDB)
- [X] Tests
- [ ] Deployment
- [ ] CLI
- [ ] General tooling
- [X] Documentation

## Types of changes
- [ ] Bug fix (generally a non-breaking change which closes an issue).
- [X] Enhancement or new feature (adds new functionality).
- [ ] Breaking change (a bug fix or enhancement which changes existing behavior).

## Checklist:
- [X] I signed an [Apache CLA](https://github.com/apache/openwhisk/blob/master/CONTRIBUTING.md).
- [X] I reviewed the [style guides](https://github.com/apache/openwhisk/blob/master/CONTRIBUTING.md#coding-standards) and followed the recommendations (Travis CI will check :).
- [X] I added tests to cover my changes.
- [X] My changes require further changes to the documentation.
- [X] I updated the documentation where necessary.

### CODE REVIEW

**Comment by bdoyle0182** [comment]
I still can't decide on naming this action config `containerConcurrency` or `instanceConcurrency`. Instance is a little more generic towards whatever the future may hold for virtualization terminology so now I find myself leaning towards that, though the current terminology is probably more self explanatory.

File: core/scheduler/src/main/scala/org/apache/openwhisk/core/scheduler/queue/MemoryQueue.scala [hunk 1]
```diff
@@ -930,8 +930,16 @@ class MemoryQueue(private val etcdClient: EtcdClient,
...
+
       getUserLimit(invocationNamespace).andThen {
-        case Success(limit) =>
+        case Success(namespaceLimit) =>
+          // extra safeguard to use namespace limit if action limit exceeds due to namespace limit being lowered
+          // by operator after action is deployed
+          val actionLimit = actionMetaData.limits.containerConcurrency
+            .map(limit =>
+              if (limit.maxConcurrentContainers > namespaceLimit) ContainerConcurrencyLimit(namespaceLimit) else limit)
+            .getOrElse(ContainerConcurrencyLimit(namespaceLimit))
```
Comment (style95) [inline core/scheduler/src/main/scala/org/apache/openwhisk/core/scheduler/queue/MemoryQueue.scala:941]: This `ContainerConcurrencyLimit` reminds me of intra-container concurrency.
Can we use `ActionConcurrency` instead? I think we need to consider changing the existing `ActionConcurrency` to something else like `IntraConcurrency` to make it more intuitive.

**Review by style95 — COMMENTED** [review]
It looks generally good to me with one comment.

File: core/scheduler/src/main/scala/org/apache/openwhisk/core/scheduler/queue/MemoryQueue.scala (hunk 1 above)
Comment (bdoyle0182) [reply core/scheduler/src/main/scala/org/apache/openwhisk/core/scheduler/queue/MemoryQueue.scala:941]: I agree but that's going to be a breaking change for the existing users of it. Is there a way we could add another new field that maps to the same field stored in the db? I guess I just don't want to have to change the api.

I could rename it to `ActionInstanceLimit`? And rename the existing Limit case class to `ActionIntraConcurrencyLimit`.

However the api would have to remain the same such that the two fields the user supplies on the api would be 
```
concurrency
instanceConcurrency
```

or I could add a third field to the api such that there's these field options on the action put
```
concurrency
intraConcurrency
instanceConcurrency
```

where if both `intraConcurrency` and `concurrency` are defined, `concurrency` is ignored and `intraConcurrency` takes precedence as the value that's used in the service. I could make it such that if you try to upload with both fields the api rejects, and any new uploads of actions with either field defined writes to `intraConcurrency` in the db and removes `concurrency` from the document if it exists.


thoughts?

Comment (style95) [reply core/scheduler/src/main/scala/org/apache/openwhisk/core/scheduler/queue/MemoryQueue.scala:941]: Yes, that sounds good to me.
So we can support both `concurrency` and `intraConcurrency` for a while and finally remove `concurrency` from API at some point.
The `instanceConcurrency` is a newly added field.

Comment (bdoyle0182) [reply core/scheduler/src/main/scala/org/apache/openwhisk/core/scheduler/queue/MemoryQueue.scala:941]: Yep. I will get to cleaning up and renaming. Will let you know when I'm ready for another look

Comment (style95) [reply core/scheduler/src/main/scala/org/apache/openwhisk/core/scheduler/queue/MemoryQueue.scala:941]: @bdoyle0182 Thank you so much.

Comment (bdoyle0182) [reply core/scheduler/src/main/scala/org/apache/openwhisk/core/scheduler/queue/MemoryQueue.scala]: @style95 so I went a little different. I changed the api fields and db fields to be:

```
concurrency
instances
```

I think this is sufficient in distinguishing the two mechanisms and they follow the same kind of verbiage for the api fields and keeps us from having to go through a whole api change process which I think can be avoided. I updated the code and documentation to refer to `concurrency` as intra concurrency including the swagger as well as the additions of this mr to `InstanceConcurrency` in code.

The `concurrency` and `instances` field go into the `limits` section of the actions document / api so it should be self explanatory that they represent a max and is explained in the documentation.

Comment (style95) [reply core/scheduler/src/main/scala/org/apache/openwhisk/core/scheduler/queue/MemoryQueue.scala]: Yeah, that would be better 👍

**Review by style95 — APPROVED** [review]
LGTM!

**Pull request merged**